sourcing/domain/service.go

```diff
@@ -0,0 +1,318 @@
+		if err != nil {
+			d.logger.Error(err)
+			lastStockError = err
+
+			continue
+		}
+		if qty > 0 {
+			availableSources[source] = qty
+		}
+	}
+
+	// if a cart is given we need to deduct the possible allocated items in the cart
+	if decoratedCart != nil {
+		allocatedSources, err := d.AllocateItems(ctx, decoratedCart)
+		if err != nil {
+			return nil, err
+		}
+
+		itemIdsWithProduct := getItemIdsWithProduct(decoratedCart, product)
+
+		for _, itemID := range itemIdsWithProduct {
+			availableSources = availableSources.Reduce(allocatedSources[itemID].AllocatedQtys)
+		}
+	}
+
+	if len(availableSources) == 0 && lastStockError != nil {
+		return availableSources, errors.Wrap(ErrNoSourceAvailable, lastStockError.Error())
+	} else if len(availableSources) == 0 {
+		return availableSources, ErrNoSourceAvailable
+	}
+
+	return availableSources, nil
+}
+
+// AllocateItems - see description in Interface
+func (d *DefaultSourcingService) AllocateItems(ctx context.Context, decoratedCart *decorator.DecoratedCart) (ItemAllocations, error) {
+	if err := d.checkConfiguration(); err != nil {
+		return nil, err
+	}
+	if decoratedCart == nil {
+		return nil, errors.New("Cart not given")
+	}
+
+	// productSourcestock holds the availablestock per product and source.
+	// During allocation the initial retrieved available stock is reduced according to used allocation
+	var productSourcestock = map[string]map[Source]int{}
+
+	if len(decoratedCart.DecoratedDeliveries) == 0 {
+		return nil, ErrNeedMoreDetailsSourceCannotBeDetected
+	}
+
+	resultItemAllocations := ItemAllocations{}
+
+	// overallError that will be returned
+	var overallError error
+	for _, delivery := range decoratedCart.DecoratedDeliveries {
+		for _, decoratedItem := range delivery.DecoratedItems {
+			var itemAllocation ItemAllocation
+			itemAllocation, productSourcestock = d.allocateItem(ctx, productSourcestock, decoratedItem, delivery.Delivery.DeliveryInfo)
+			resultItemAllocations[ItemID(decoratedItem.Item.ID)] = itemAllocation
+		}
+	}
+
+	return resultItemAllocations, overallError
+}
+
+func (d *DefaultSourcingService) checkConfiguration() error {
+	if d.availableSourcesProvider == nil {
+		d.logger.Error("no Source Provider bound")
+		return errors.New("no Source Provider bound")
+	}
+	if d.stockProvider == nil {
+		d.logger.Error("no Stock Provider bound")
+		return errors.New("no Stock Provider bound")
+	}
+
+	return nil
+}
+
+func getItemIdsWithProduct(dc *decorator.DecoratedCart, product domain.BasicProduct) []ItemID {
+	var result []ItemID
+	for _, di := range dc.GetAllDecoratedItems() {
+		if di.Product.GetIdentifier() == product.GetIdentifier() {
+			result = append(result, ItemID(di.Item.ID))
+		}
+	}
+	return result
+}
+
+// allocateItem returns the itemAllocation and the remaining stock for the given item.
+// The passed productSourcestock is used - and the remaining productSourcestock is returned. In case a source is not yet goven in productSourcestock it will be fetched
```

Comment:
```suggestion
// The passed productSourcestock is used - and the remaining productSourcestock is returned. In case a source is not yet given in productSourcestock it will be fetched
```

Reply:
Fixed